src/conversions/bytes.rs

```diff
@@ -0,0 +1,161 @@
+#![cfg(feature = "bytes")]
+
+//! Conversions to and from [bytes](https://docs.rs/bytes/latest/bytes/)'s [`Bytes`] and
+//! [`BytesMut`] types.
+//!
+//! This is useful for efficiently converting Python's `bytes` and `bytearray` types efficiently.
+//!
+//! # Setup
+//!
+//! To use this feature, add in your **`Cargo.toml`**:
+//!
+//! ```toml
+//! [dependencies]
+#![doc = concat!("pyo3 = { version = \"", env!("CARGO_PKG_VERSION"),  "\", features = [\"bytes\"] }")]
+//! bytes = "1.10"
+//!
+//! Note that you must use compatible versions of bytes and PyO3.
+//!
+//! # Example
+//!
+//! Rust code to create functions which return `Bytes` or take `Bytes` as arguements:
+//!
+//! ```rust,no_run
+//! use pyo3::prelude::*;
+//!
+//! #[pyfunction]
+//! fn get_message_bytes() -> Bytes {
+//!     Bytes::from(b"Hello Python!".to_vec())
+//! }
+//!
+//! #[pyfunction]
+//! fn num_bytes(bytes: Bytes) -> usize {
+//!     bytes.len()
+//! }
+//!
+//! #[pymodule]
+//! fn my_module(m: &Bound<'_, PyModule>) -> PyResult<()> {
+//!     m.add_function(wrap_pyfunction!(get_message_bytes, m)?)?;
+//!     m.add_function(wrap_pyfunction!(num_bytes, m)?)?;
+//!     Ok(())
+//! }
+//! ```
+//!
+//! Python code that calls these functions:
+//!
+//! ```python
+//! from my_module import get_message_bytes, num_bytes
+//!
+//! message = get_message_bytes()
+//! assert message == b"Hello Python!"
+//!
+//! size = num_bytes(message)
+//! assert size == 13
+//! ```
+use bytes::{Bytes, BytesMut};
+
+use crate::conversion::IntoPyObject;
+use crate::exceptions::PyTypeError;
+use crate::instance::Bound;
+use crate::types::any::PyAnyMethods;
+use crate::types::{PyByteArray, PyByteArrayMethods, PyBytes, PyBytesMethods};
+use crate::{FromPyObject, PyAny, PyErr, PyResult, Python};
+
+impl FromPyObject<'_> for Bytes {
+    fn extract_bound(ob: &Bound<'_, PyAny>) -> PyResult<Self> {
+        if let Ok(bytes) = ob.downcast::<PyBytes>() {
+            Ok(Bytes::from((*bytes).as_bytes().to_vec()))
+        } else if let Ok(bytearray) = ob.downcast::<PyByteArray>() {
+            Ok(Bytes::from((*bytearray).to_vec()))
+        } else {
+            Err(PyTypeError::new_err("expected bytes or bytearray"))
+        }
```

Comment:
As discussed elsewhere I think we can use `from_owner` on a buffer; separately I can fix #4632 which I think is the reason why `pyo3_bytes` has a wrapper.

We need to be careful about mutable buffers; as `PyByteArray` is mutable it is probably unsound to directly reference it. If we choose to support it (or other potentially mutable objects) we should add tests to demonstrate that having a readonly buffer prevents modification of the original object. I think this might be implementation dependent, so in the end it might be that only `PyBytes` is sound.